libdrgn/drgn.h

```diff
@@ -927,9 +938,81 @@ struct drgn_error *drgn_program_find_symbols_by_name(struct drgn_program *prog,
  */
 struct drgn_error *drgn_program_find_symbols_by_address(struct drgn_program *prog,
 							uint64_t address,
+							struct drgn_module *module,
 							struct drgn_symbol ***syms_ret,
 							size_t *count_ret);
 
+/** Flags for @ref drgn_program_find_symbols() */
+enum drgn_find_symbol_flags {
+	/** Find symbols whose name matches the name argument */
+	DRGN_FIND_SYM_NAME = 1 << 0,
+	/** Find symbols whose address matches the addr argument */
+	DRGN_FIND_SYM_ADDR = 1 << 1,
+	/** Find only one symbol */
+	DRGN_FIND_SYM_ONE = 1 << 2,
+};
+
+/** Result builder for @ref drgn_program_find_symbols() */
+struct drgn_symbol_result_builder;
+
+/**
+ * Add or set the return value for a symbol search
+ *
+ * Symbol finders should call this with each symbol search result. If the symbol
+ * search was @ref DRGN_FIND_SYM_ONE, then only the most recent symbol added to
+ * the builder will be returned. Otherwise, all symbols added to the builder are
+ * returned.
+ */
+struct drgn_error *
+drgn_symbol_result_builder_add(struct drgn_symbol_result_builder *builder,
+			       struct drgn_symbol *symbol);
+
+size_t drgn_symbol_result_builder_count(const struct drgn_symbol_result_builder *builder);
+
+/**
+ * Callback for finding one or more symbols.
+ *
+ * The callback should perform a symbol lookup based on the flags given in @a
+ * flags. When multiple flags are provided, the effect should be treated as a
+ * logical AND. Symbol results should be added to the result builder @a builder,
+ * via @ref drgn_symbol_result_builder_add(). When @ref DRGN_FIND_SYM_ONE is
+ * set, then the finding function should only return the single best symbol
+ * result, and short-circuit return.
+ *
+ * When @a module is not NULL, it must be used to limit the result to only
+ * symbols from the specified module.
+ *
+ * When no symbol is found, simply do not add any result to the builder. No
+ * error should be returned in this case.
+ *
+ * @param[in] name Name of the symbol to match
+ * @param[in] addr Address of the symbol to match
+ * @param[in] module Drgn module to limit search to
+ * @param[in] flags Flags indicating the desired behavior of the search
+ * @param[in] arg Argument passed to @ref drgn_program_add_symbol_finder().
+ * @param[in] builder Used to build the resulting symbol output
+ */
+typedef struct drgn_error *
+(*drgn_find_symbol_fn)(const char *name, uint64_t addr,
+		       struct drgn_module *module,
```

Comment:
From our discussion yesterday, we will be removing `struct drgn_module *` from the API. Marking this comment here so I don't forget to actually do it!